pkg/pillar/cmd/zedagent/handlecertconfig.go

```diff
@@ -127,70 +142,181 @@ func unpublishControllerCert(ctx *getconfigContext, key string) {
 func handleEveNodeCertModify(ctxArg interface{}, key string,
 	configArg interface{}) {
 
-	// XXX TBD
+	ctx := ctxArg.(*zedagentContext)
 	status := configArg.(types.EveNodeCert)
 	log.Infof("handleEveNodeCertModify for %s", status.Key())
+	triggerEveNodeCertEvent(ctx)
 	return
 }
 
 func handleEveNodeCertDelete(ctxArg interface{}, key string,
 	configArg interface{}) {
 
-	// XXX TBD
+	ctx := ctxArg.(*zedagentContext)
 	status := configArg.(types.EveNodeCert)
 	log.Infof("handleEveNodeCertDelete for %s", status.Key())
+	triggerEveNodeCertEvent(ctx)
 	return
 }
 
 // Run a task certificate post task, on change trigger
-func certsTask(ctx *zedagentContext, triggerCerts <-chan struct{}) {
-	iteration := 0
+func controllerCertsTask(ctx *zedagentContext, triggerCerts <-chan struct{}) {
+
 	log.Infoln("starting certs publish task")
-	publishCerts(ctx, iteration)
+	getCertsFromController(ctx)
 
 	// Run a periodic timer so we always update StillRunning
 	stillRunning := time.NewTicker(25 * time.Second)
-	agentlog.StillRunning(agentName+"certs", warningTime, errorTime)
+	agentlog.StillRunning(agentName+"ccerts", warningTime, errorTime)
 
 	for {
 		select {
 		case <-triggerCerts:
 			start := time.Now()
-			iteration++
-			publishCerts(ctx, iteration)
-			pubsub.CheckMaxTimeTopic(agentName+"certs", "publishCerts", start,
+			getCertsFromController(ctx)
+			pubsub.CheckMaxTimeTopic(agentName+"ccerts", "publishCerts", start,
 				warningTime, errorTime)
 
 		case <-stillRunning.C:
 		}
-		agentlog.StillRunning(agentName+"certs", warningTime, errorTime)
+		agentlog.StillRunning(agentName+"ccerts", warningTime, errorTime)
 	}
 }
 
 // prepare the certs list proto message
-func publishCerts(ctx *zedagentContext, iteration int) {
+func getCertsFromController(ctx *zedagentContext) bool {
+	certURL := zedcloud.URLPathString(serverNameAndPort,
+		zedcloudCtx.V2API, nilUUID, "certs")
+
+	// not V2API
+	if !zedcloud.UseV2API() {
+		return false
+	}
+
+	reqID := zcdevUUID.String()
+	zedcloud.RemoveDeferred("ccerts:" + reqID)
```

Comment:
I would be more maintainable if you defined a local deferKey := "ccerts:"+reqID and use it here plus in the SetDeferred below.
Non-obvious things would happen if the two ever get out of synch due to some future code changes.

But WAIT, this is a get and not a POST, right? So there is no deferred processing needed. Just log the error and presumably a trigger will happen again.
When does the hash used by the trigger get updated? For it to trigger again you should wait to update the hash of the controller certs until you have received the certs - I haven't checked if that is already the case in this code.

Reply:
Checked and that is indeed the case. So you can remove the RemoveDeferred and SetDeferred for this get.

Reply:
removed.